src/kemal/middleware/filters.cr

```diff
@@ -21,54 +21,49 @@ module Kemal::Middleware
-    def filter_for_path_type_defined?(verb, path, type)
-      lookup = @tree.find radix_path(verb, path, type)
-      lookup.found? && lookup.payload.is_a? Block
-    end
-
     # :nodoc: This shouldn't be called directly, it's not private because I need to call it for testing purpose since I can't call the macros in the spec.
     # It adds the block for the corresponding verb/path/type combination to the tree.
     def _add_route_filter(verb, path, type, &block : HTTP::Server::Context -> _)
-      node = radix_path(verb, path, type)
-      @tree.add node, Block.new &block
+      lookup = lookup_filters_for_path_type(verb, path, type)
+      if lookup.found? && lookup.payload.is_a?(Array(Block))
+        (lookup.payload as Array(Block)) << Block.new(&block)
+      else
+        @tree.add radix_path(verb, path, type), [Block.new(&block)]
+      end
     end
 
     # This can be called directly but it's simpler to just use the macros, it will check if another filter is not already defined for this verb/path/type and proceed to call `add_route_filter`
     def before(verb, path = "*", &block : HTTP::Server::Context -> _)
-      raise Kemal::Middleware::Filter::BeforeFilterAlreadyDefinedException.new(verb, path) if filter_for_path_type_defined?(verb, path, :before)
       _add_route_filter verb, path, :before, &block
     end
 
     # This can be called directly but it's simpler to just use the macros, it will check if another filter is not already defined for this verb/path/type and proceed to call `add_route_filter`
     def after(verb, path = "*", &block : HTTP::Server::Context -> _)
-      raise Kemal::Middleware::Filter::AfterFilterAlreadyDefinedException.new(verb, path) if filter_for_path_type_defined?(verb, path, :after)
       _add_route_filter verb, path, :after, &block
     end
 
     # This will fetch the block for the verb/path/type from the tree and call it.
     private def call_block_for_path_type(verb, path, type, context)
-      lookup = @tree.find radix_path(verb, path, type)
-      if lookup.found? && lookup.payload.is_a? Block
-        block = lookup.payload as Block
-        block.block.call(context)
+      lookup = lookup_filters_for_path_type(verb, path, type)
+      if lookup.found? && lookup.payload.is_a? Array(Block)
+        ary = lookup.payload as Array(Block)
+        ary.each { |block| block.block.call(context) }
       end
     end
 
-    private def radix_path(verb, path, type : Symbol)
-      "#{type}/#{verb}/#{path}"
+    # This checks is filter is already defined for the verb/path/type combination
+    private def filter_for_path_type_defined?(verb, path, type)
+      lookup = @tree.find radix_path(verb, path, type)
+      lookup.found? && lookup.payload.is_a? Block
     end
 
-    class BeforeFilterAlreadyDefinedException < Exception
-      def initialize(verb, path)
-        super "A before-filter is already defined for path: '#{verb}:#{path}'."
-      end
+    # This returns a lookup for verb/path/type
```

Comment:
it returns ONE lookup which contains a paylof of ONE array hence the name :smile: